src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableUpcallHandler.java

```diff
@@ -155,11 +134,43 @@ private static void checkPrimitive(MethodType type) {
     }
 
     private static MethodHandle specializedBindingHandle(MethodHandle target, CallingSequence callingSequence,
-                                                         Class<?> llReturn, long allocationSize) {
+                                                         Class<?> llReturn, ABIDescriptor abi) {
         MethodType highLevelType = callingSequence.methodType();
 
         MethodHandle specializedHandle = target; // initial
 
+        // we handle returns first since IMR adds an extra parameter that needs to be specialized as well
+        if (llReturn != void.class || callingSequence.needsReturnBuffer()) {
```

Comment:
Note that the old code for handling return values is just moved from below the code that handles arguments, since the return buffer argument also needs to be handle along with the rest of arguments, and it is added here.